torchvision/datasets/folder.py

```diff
@@ -32,20 +32,32 @@ def is_image_file(filename):
     return has_file_allowed_extension(filename, IMG_EXTENSIONS)
 
 
-def make_dataset(dir, class_to_idx, extensions):
+def make_dataset(dir, class_to_idx, extensions, is_valid_file=None):
     images = []
     dir = os.path.expanduser(dir)
+    if extensions is None and is_valid_file is None:
+        raise ValueError("Both extensions and is_valid_file cannot be None")
+    if extensions is not None and is_valid_file is not None:
+        raise ValueError("One of the extensions and is_valid_file should be None")
+
     for target in sorted(class_to_idx.keys()):
         d = os.path.join(dir, target)
         if not os.path.isdir(d):
             continue
-
-        for root, _, fnames in sorted(os.walk(d)):
-            for fname in sorted(fnames):
-                if has_file_allowed_extension(fname, extensions):
+        if extensions is not None:
```

Comment:
I think we can avoid duplicating this codepath by having a default `is_valid_file` that defaults to `has_file_allowed_extension`. Something like
```python
if is_valid_file is None:
    is_valid_file = lambda x: has_file_allowed_extension(x, extensions)
```

and use `is_valid_file` everywhere in the function

Reply:
done